pubsub/pubsub.go

```diff
@@ -0,0 +1,272 @@
+	}
+	t := &Topic{
+		driver:  d,
+		batcher: b,
+	}
+	return t
+}
+
+// Subscription receives published messages.
+type Subscription struct {
+	driver driver.Subscription
+
+	// ackBatcher makes batches of acks and sends them to the server.
+	ackBatcher *bundler.Bundler
+
+	// q is the local queue of messages downloaded from the server.
+	sem chan struct{}
+	q   []*Message
+}
+
+// SubscriptionOptions contains configuration for Subscriptions.
+type SubscriptionOptions struct {
+	// AckDelay tells the max duration to wait before sending the next batch
+	// of acknowledgements back to the server.
+	AckDelay time.Duration
+
+	// AckBatchSize is the maximum number of acks that should be sent to
+	// the server in a batch.
+	AckBatchSize int
+
+	// AckDeadline tells how long the server should wait before assuming a
+	// received message has failed to be processed.
+	AckDeadline time.Duration
+}
+
+// AckDelayDefault is the value for SubscriptionOptions.AckDelay if it is not set.
+const AckDelayDefault = time.Millisecond
+
+// Receive receives and returns the next message from the Subscription's queue,
+// blocking and polling if none are available. This method can be called
+// concurrently from multiple goroutines. The Ack() method of the returned
+// Message has to be called once the message has been processed, to prevent it
+// from being received again.
+func (s *Subscription) Receive(ctx context.Context) (*Message, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-s.sem:
+	}
+	defer func() {
+		s.sem <- struct{}{}
+	}()
+	if len(s.q) == 0 {
+		if err := s.getNextBatch(ctx); err != nil {
+			return nil, err
+		}
+	}
+	m := s.q[0]
+	s.q = s.q[1:]
+	return m, nil
+}
+
+// getNextBatch gets the next batch of messages from the server.
+func (s *Subscription) getNextBatch(ctx context.Context) error {
+	msgs, err := s.driver.ReceiveBatch(ctx)
+	if err != nil {
+		return err
+	}
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		if len(msgs) == 0 {
+			return errors.New("subscription driver bug: received empty batch")
+		}
+		s.q = make([]*Message, len(msgs))
+		for i, m := range msgs {
+			s.q[i] = &Message{
+				Body:       m.Body,
+				Attributes: m.Attributes,
+				ackID:      m.AckID,
+				sub:        s,
+			}
+		}
+	}
+	return nil
+}
+
+// Close flushes pending ack sends and disconnects the Subscription.
+func (s *Subscription) Close() error {
+	s.ackBatcher.Flush()
+	return s.driver.Close()
+}
+
+// NewSubscription creates a Subscription from a driver.Subscription and opts to
+// tune sending and receiving of acks and messages. Behind the scenes,
+// NewSubscription spins up a goroutine to gather acks into batches and
+// periodically send them to the server.
+func NewSubscription(ctx context.Context, d driver.Subscription, opts SubscriptionOptions) *Subscription {
+	handler := func(item interface{}) {
+		mecs := item.([]msgErrChan)
+		var ids []driver.AckID
+		for _, mec := range mecs {
```

Comment:
same comment as above

Reply:
I don't have enough context here (apologies, I am failing really hard at following this review through the PR UI), but IIUC @vangent's suggestion is to do something like `ids[i] = mecs[i].msg.ackID`. I don't have a strong opinion here. Most of the way around the inefficiencies are to set the capacity correctly anyway.

Reply:
I can't find anything in the readability guidelines about this, so I'll waive my objection if Robert insists. I'll just reiterate the flaws:
- More code on the page, so harder to read.
- Potential bug: writing a different slice in `make([]T, len(s))` and `for i, x := range s` (which also means extra effort on the reader's part to check for this).
- Overallocation (and complex handling of the loop index) when only some of the results are used. E.g. compare:
```
a := make([]T, len(b))
for i, x := range b {
   a[i] = x
}
```
with
```
a := make([]T, len(b))
for i := 0; i < len(b); {
   x := b[i]
   if filter(x) {
      a[i] = x
      i++
  }
}
```
Whereas if append is used, the loops look almost identical.

That's why I call this a premature optimization: it's faster, sure, but it makes the code worse overall, so it should only be done if performance is shown to matter.

Reply:
I found some guidance from Ian Lance Taylor, which I'll quote:

> I'm OK with [the loop that uses `a[i] = x`] when it really is that
simple.  As soon as there are several statements, or a conditional, in
the loop body, I think that using append is preferable stylistically,
as it will clearly continue to work if the code changes such that
there are cases where the value is not added to the slice.

So I'm fine with using the index in this simple case, but we should be wary in more complex loops.

Reply:
Okay, I changed all of these loops to use `append`.